File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SwapPairs.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

public class SwapPairs {
    public ListNode swapPairs(ListNode head) {
        // 递归终止条件：没有节点或只有一个节点
        if (head == null || head.next == null) return head;

        ListNode firstNode = head;
        ListNode secondNode = head.next;
        ListNode remainingNodes = secondNode.next;

        // 递归处理后续节点
        firstNode.next = swapPairs(remainingNodes);

        // 交换当前两个节点
        secondNode.next = firstNode;

        return secondNode;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/swapPairs.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun swapPairs(head: ListNode?): ListNode? {
    // 递归终止条件：没有节点或只有一个节点
    if (head?.next == null) return head

    val firstNode = head
    val secondNode = firstNode.next!!
    val remainingNodes = secondNode.next

    // 递归处理后续节点
    firstNode.next = swapPairs(remainingNodes)

    // 交换当前两个节点
    secondNode.next = firstNode

    return secondNode
}
